api/src/services/items.ts

```diff
@@ -481,43 +488,39 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
 
 					const snapshots = await itemsService.readMany(keys);
 
-					const revisionRecords: {
-						activity: PrimaryKey;
-						collection: string;
-						item: PrimaryKey;
-						data: string;
-						delta: string;
-					}[] = [];
-
-					for (let i = 0; i < activityPrimaryKeys.length; i++) {
-						revisionRecords.push({
-							activity: activityPrimaryKeys[i],
-							collection: this.collection,
-							item: keys[i],
-							data: snapshots && Array.isArray(snapshots) ? JSON.stringify(snapshots[i]) : JSON.stringify(snapshots),
-							delta: await payloadService.prepareDelta(payloadWithTypeCasting),
-						});
-					}
-
-					for (let i = 0; i < revisionRecords.length; i++) {
-						const revisionID = (
-							await trx.insert(revisionRecords[i]).into('directus_revisions').returning('id')
-						)[0] as number;
+					const revisionsService = new RevisionsService({
+						knex: trx,
+						schema: this.schema,
+					});
 
+					const revision = await revisionsService.createMany(
+						await Promise.all(
+							activity.map(async (activity, index) => ({
+								activity: activity,
+								collection: this.collection,
+								item: keys[index],
+								data:
+									snapshots && Array.isArray(snapshots) ? JSON.stringify(snapshots[index]) : JSON.stringify(snapshots),
+								delta: await payloadService.prepareDelta(payloadWithTypeCasting),
+							}))
+						)
+					);
```

Comment:
@rijkvanzanten I kept the logic for revision creation same as before (loop).
However I was asking myself: How sure can we be, that the order of activity entries and key entries will always be the same?
As safer approach might be to fetch the information about the activity entries which have just been created and use those to map it to the correct keys...